Tasks/Npm/util.ts

```diff
@@ -0,0 +1,158 @@
+    let collectionUrl = url.parse(tl.getVariable('System.TeamFoundationCollectionUri'));
+
+    if (collectionUrl.hostname.toUpperCase().endsWith('.VISUALSTUDIO.COM')) {
+        let hostParts = collectionUrl.hostname.split('.');
+        let packagingHostName = hostParts[0] + '.pkgs.visualstudio.com';
+        collectionUrl.hostname = packagingHostName;
+        // remove the host property so it doesn't override the hostname property for url.format
+        delete collectionUrl.host;
+    }
+
+    return Q(url.format(collectionUrl));
+}
+
+export function getTempNpmrcPath(): string {
+    let tempUserNpmrcPath: string = path.join(getTempPath(), `${tl.getVariable('Build.BuildId')}.npmrc`);
+
+    return tempUserNpmrcPath;
+}
+
+export function getTempPath(): string {
+    let tempNpmrcDir
+        = tl.getVariable('Agent.BuildDirectory')
+        || tl.getVariable('Agent.ReleaseDirectory')
+        || process.cwd();
+    let tempPath = path.join(tempNpmrcDir, 'npm');
+    if (tl.exist(tempPath) === false) {
+        tl.mkdirP(tempPath);
+    }
+
+    return tempPath;
+}
+
+function copyFile(src: string, dst: string): void {
+    let content = fs.readFileSync(src);
+    fs.writeFileSync(dst, content);
+}
+
+export function saveFile(file: string): void {
+    if (file && tl.exist(file)) {
+        let tempPath = getTempPath();
+        let baseName = path.basename(file);
+        let destination = path.join(tempPath, baseName);
+
+        tl.debug(tl.loc('SavingFile', file));
+        copyFile(file, destination);
+    }
+}
+
+export function restoreFile(file: string): void {
+    if (file) {
+        let tempPath = getTempPath();
+        let baseName = path.basename(file);
+        let source = path.join(tempPath, baseName);
+
+        if (tl.exist(source)) {
+            tl.debug(tl.loc('RestoringFile', file));
+            copyFile(source, file);
+            tl.rmRF(source);
+        }
+    }
+}
+
+export function getSystemAccessToken(): string {
+    let auth = tl.getEndpointAuthorization('SYSTEMVSSCONNECTION', false);
+    if (auth.scheme === 'OAuth') {
+        tl.debug(tl.loc('FoundBuildCredentials'));
+        return auth.parameters['AccessToken'];
+    } else {
+        tl.warning(tl.loc('NoBuildCredentials'));
+    }
+
+    return undefined;
+}
+
+export function toNerfDart(uri: string): string {
+    var parsed = url.parse(uri);
+    delete parsed.protocol;
+    delete parsed.auth;
+    delete parsed.query;
+    delete parsed.search;
+    delete parsed.hash;
+
+    return url.resolve(url.format(parsed), '.');
+}
+
+export async function getFeedRegistryUrl(feedId: string): Promise<string> {
+    const apiVersion = '3.0-preview.1';
+    const area = 'npm';
+    const locationId = 'D9B75B07-F1D9-4A67-AAA6-A4D9E66B3352';
+
+    let accessToken = getSystemAccessToken();
+    let credentialHandler = vsts.getBearerHandler(accessToken);
+    let collectionUrl = await getPackagingCollectionUrl();
```

Comment:
Just as it is, it would be hard to test it in DevFabric. Adding some test hooks for the URL could be useful.